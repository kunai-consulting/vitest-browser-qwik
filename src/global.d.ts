import type { SymbolMapperFn } from "@qwik.dev/core/optimizer";

declare global {
  var qwikSymbolMapper: SymbolMapperFn | undefined;
}
