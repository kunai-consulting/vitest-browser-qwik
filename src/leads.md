- First: support CSR then we can support SSR (unless its easier to support both)
- Figure out, do we need a Container type?
- What the heck is createConcurrentRoot? Do we need it?
- WTF is act? Do we need it?

Hints:
- Look through Qwik's existing types.


Dev entry that Qwik has

```tsx
/*
 * WHAT IS THIS FILE?
 *
 * Development entry point using only client-side modules:
 * - Do not use this mode in production!
 * - No SSR
 * - No portion of the application is pre-rendered on the server.
 * - All of the application is running eagerly in the browser.
 * - More code is transferred to the browser than in SSR mode.
 * - Optimizer/Serialization/Deserialization code is not exercised!
 */
import { render, type RenderOptions } from "@builder.io/qwik";
import Root from "./root";

export default function (opts: RenderOptions) {
  return render(document, <Root />, opts);
}
```