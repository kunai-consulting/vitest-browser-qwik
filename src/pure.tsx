import type { Component, JSXOutput, RenderOptions } from "@builder.io/qwik";
import type { Locator, LocatorSelectors } from "@vitest/browser/context";
import {
	debug,
	getElementLocatorSelectors,
	type PrettyDOMOptions,
} from "@vitest/browser/utils";

export interface Options extends RenderOptions {
	container?: HTMLElement;
	baseElement?: HTMLElement;
	wrapper?: Component;
}

export type DebugFn = (
	baseElement?: HTMLElement | HTMLElement[],
	maxLength?: number,
	options?: PrettyDOMOptions,
) => void;

export type RenderResult = {
	asFragment: () => DocumentFragment;
	container: HTMLElement;
	baseElement: HTMLElement;
	debug: DebugFn;
	unmount: () => void;
};

export type ComponentRef = {
	container: HTMLElement;
	componentCleanup: () => void;
};

// export interface RenderResult extends LocatorSelectors {
// 	container: HTMLElement;
// 	baseElement: HTMLElement;
// 	debug: (
// 		el?: HTMLElement | HTMLElement[] | Locator | Locator[],
// 		maxLength?: number,
// 		options?: PrettyDOMOptions,
// 	) => void;
// 	unmount: () => void;
// 	rerender: (ui: React.ReactNode) => void;
// 	asFragment: () => DocumentFragment;
// }

export interface ComponentRenderOptions {
	container?: HTMLElement;
	baseElement?: HTMLElement;
	wrapper?: JSXOutput;
}

// Ideally we'd just use a WeakMap where containers are keys and roots are values.
// We use two variables so that we can bail out in constant time when we render with a new container (most common use case)
const mountedContainers = new Set<Container>();
const mountedRootEntries: {
	container: Container;
	root: ReturnType<typeof createConcurrentRoot>;
}[] = [];

export function render(
	ui: JSXOutput,
	{
		container,
		baseElement,
		wrapper: WrapperComponent,
	}: ComponentRenderOptions = {},
): RenderResult {
	if (!baseElement) {
		// default to document.body instead of documentElement to avoid output of potentially-large
		// head elements (such as JSS style blocks) in debug output
		baseElement = document.body;
	}

	if (!container) {
		container = baseElement.appendChild(document.createElement("div"));
	}

	let root: QwikRoot;

	if (!mountedContainers.has(container)) {
		root = createConcurrentRoot(container);

		mountedRootEntries.push({ container, root });
		// we'll add it to the mounted containers regardless of whether it's actually
		// added to document.body so the cleanup method works regardless of whether
		// they're passing us a custom container or not.
		mountedContainers.add(container);
	} else {
		mountedRootEntries.forEach((rootEntry) => {
			// Else is unreachable since `mountedContainers` has the `container`.
			// Only reachable if one would accidentally add the container to `mountedContainers` but not the root to `mountedRootEntries`
			/* istanbul ignore else */
			if (rootEntry.container === container) {
				root = rootEntry.root;
			}
		});
	}

	return {
		container,
		baseElement,
		debug: (el, maxLength, options) => debug(el, maxLength, options),
		unmount: () => {
			root.unmount();
		},
		asFragment: () => {
			return document
				.createRange()
				.createContextualFragment(container.innerHTML);
		},
		...getElementLocatorSelectors(baseElement),
	};
}

export interface RenderHookOptions<Props> extends ComponentRenderOptions {
	/**
	 * The argument passed to the renderHook callback. Can be useful if you plan
	 * to use the rerender utility to change the values passed to your hook.
	 */
	initialProps?: Props | undefined;
}

export interface RenderHookResult<Result, Props> {
	/**
	 * This is a stable reference to the latest value returned by your renderHook
	 * callback
	 */
	result: {
		/**
		 * The value returned by your renderHook callback
		 */
		current: Result;
	};
	/**
	 * Unmounts the test component. This is useful for when you need to test
	 * any cleanup your useEffects have.
	 */
	unmount: () => void;
}

export function renderHook<Props, Result>(
	renderCallback: (initialProps?: Props) => Result,
	options: RenderHookOptions<Props> = {},
): RenderHookResult<Result, Props> {
	const { initialProps, ...renderOptions } = options;

	return { result, unmount };
}

export function cleanup(): void {
	mountedRootEntries.forEach(({ root, container }) => {
		root.unmount();
		if (container.parentNode === document.body) {
			document.body.removeChild(container);
		}
	});
	mountedRootEntries.length = 0;
	mountedContainers.clear();
}

interface QwikRoot {
	render: (element: JSXOutput) => void;
	unmount: () => void;
}

function createConcurrentRoot(container: HTMLElement): QwikRoot {
	// I think this is the render function for Qwik version?
	const root = ReactDOMClient.createRoot(container);

	return {
		render(element: JSXOutput) {
			root.render(element);
		},
		unmount() {
			root.unmount();
		},
	};
}

export interface RenderConfiguration {
	reactStrictMode: boolean;
}

const config: RenderConfiguration = {
	reactStrictMode: false,
};

// can prob be removed
function strictModeIfNeeded(innerElement: React.ReactNode) {
	return config.reactStrictMode
		? React.createElement(React.StrictMode, null, innerElement)
		: innerElement;
}

// maybe removed?
function wrapUiIfNeeded(
	innerElement: React.ReactNode,
	wrapperComponent?: React.JSXElementConstructor<{
		children: React.ReactNode;
	}>,
) {
	return wrapperComponent
		? React.createElement(wrapperComponent, null, innerElement)
		: innerElement;
}

export function configure(customConfig: Partial<RenderConfiguration>): void {
	Object.assign(config, customConfig);
}
