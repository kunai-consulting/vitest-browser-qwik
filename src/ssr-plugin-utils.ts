import { dirname, relative, resolve } from "node:path";
import type { Component } from "@builder.io/qwik";
import type {
	BindingIdentifier,
	CallExpression,
	ExpressionStatement,
	FunctionType,
	ImportDeclaration,
	ImportDefaultSpecifier,
	ImportSpecifier,
	JSXAttribute,
	JSXAttributeItem,
	JSXElement,
	JSXExpressionContainer,
	Node,
	Function as OxcFunction,
	Span,
	VariableDeclarator,
} from "@oxc-project/types";
import { ResolverFactory } from "oxc-resolver";
import type { BrowserCommandContext } from "vitest/node";

const resolver = new ResolverFactory({
	extensions: [".tsx", ".ts", ".jsx", ".js"],
});

export function isFunction(node: Node): node is OxcFunction {
	const functionTypes: FunctionType[] = [
		"FunctionDeclaration",
		"FunctionExpression",
		"TSDeclareFunction",
		"TSEmptyBodyFunctionExpression",
	];
	return functionTypes.includes(node.type as FunctionType);
}

export function isCallExpression(node: Node): node is CallExpression {
	return node.type === "CallExpression";
}

export function isImportDeclaration(node: Node): node is ImportDeclaration {
	return node.type === "ImportDeclaration";
}

export function isVariableDeclarator(node: Node): node is VariableDeclarator {
	return node.type === "VariableDeclarator";
}

export function isExpressionStatement(node: Node): node is ExpressionStatement {
	return node.type === "ExpressionStatement";
}

export function isJSXElement(node: Node): node is JSXElement {
	return node.type === "JSXElement";
}

export function isJSXExpressionContainer(
	node: Node,
): node is JSXExpressionContainer {
	return node.type === "JSXExpressionContainer";
}

export function traverseChildren(
	node: Node,
	callback: (child: Node) => boolean | undefined,
): boolean {
	for (const key in node) {
		const child = (node as unknown as Record<string, unknown>)[key];
		if (Array.isArray(child)) {
			for (const item of child) {
				if (item && typeof item === "object" && callback(item as Node)) {
					return true;
				}
			}
		} else if (child && typeof child === "object" && callback(child as Node)) {
			return true;
		}
	}
	return false;
}

export function hasRenderSSRCallInAST(ast: Node, code: string): boolean {
	const renderSSRIdentifiers = new Set<string>(["renderSSR"]);
	let hasRenderSSRCallInCode = false;

	function walkForDetection(node: Node): boolean {
		if (!node || typeof node !== "object") return false;

		if (isImportDeclaration(node) && node.source?.value && node.specifiers) {
			for (const spec of node.specifiers) {
				if (spec.type === "ImportSpecifier") {
					const importSpec = spec as ImportSpecifier;
					if (
						importSpec.imported.type === "Identifier" &&
						importSpec.imported.name === "renderSSR"
					) {
						renderSSRIdentifiers.add(importSpec.local.name);
					}
				} else if (spec.type === "ImportDefaultSpecifier") {
					const defaultSpec = spec as ImportDefaultSpecifier;
					if (defaultSpec.local.name.toLowerCase().includes("renderssr")) {
						renderSSRIdentifiers.add(defaultSpec.local.name);
					}
				}
			}
		}

		if (isFunction(node) && node.id?.name === "renderSSR") {
			renderSSRIdentifiers.add("renderSSR");
		}

		if (isVariableDeclarator(node)) {
			if (
				node.id.type === "Identifier" &&
				node.init?.type === "Identifier" &&
				renderSSRIdentifiers.has(node.init.name)
			) {
				const bindingId = node.id as BindingIdentifier;
				renderSSRIdentifiers.add(bindingId.name);
			}
		}

		if (
			isCallExpression(node) &&
			node.callee.type === "Identifier" &&
			renderSSRIdentifiers.has(node.callee.name)
		) {
			hasRenderSSRCallInCode = true;
			return true;
		}

		return traverseChildren(node, walkForDetection);
	}

	walkForDetection(ast);

	return hasRenderSSRCallInCode || code.includes("renderSSR(");
}

export function extractPropsFromJSX(
	attributes: JSXAttributeItem[],
	sourceCode: string,
): Record<string, string> {
	const props: Record<string, string> = {};

	for (const attr of attributes) {
		if (attr.type !== "JSXAttribute") continue;

		const jsxAttr = attr as JSXAttribute;
		if (jsxAttr.name.type !== "JSXIdentifier" || !jsxAttr.value) continue;

		const propName = jsxAttr.name.name;

		if (
			isJSXExpressionContainer(jsxAttr.value) &&
			jsxAttr.value.expression.type !== "JSXEmptyExpression"
		) {
			const exprSpan = jsxAttr.value.expression as Node & Span;
			const expressionCode = sourceCode.slice(exprSpan.start, exprSpan.end);
			props[propName] = expressionCode;
		} else if (jsxAttr.value.type === "Literal") {
			const literal = jsxAttr.value as { value: unknown };
			props[propName] = JSON.stringify(literal.value);
		}
	}

	return props;
}

export function isTestFile(id: string): boolean {
	return id.includes(".test.") || id.includes(".spec.");
}

function fallbackResolveComponentPath(
	importPath: string,
	testFileId: string,
): string {
	if (!importPath.startsWith(".")) {
		return importPath.endsWith(".tsx") || importPath.endsWith(".ts")
			? importPath
			: `${importPath}.tsx`;
	}

	const testFileDir = dirname(testFileId);
	const resolvedPath = resolve(testFileDir, importPath);
	const projectRoot = process.cwd();
	let componentPath = `./${relative(projectRoot, resolvedPath)}`;

	if (!componentPath.endsWith(".tsx") && !componentPath.endsWith(".ts")) {
		componentPath += ".tsx";
	}

	return componentPath;
}

export function resolveComponentPath(
	importPath: string,
	testFileId: string,
): string {
	const testFileDir = dirname(testFileId);
	const result = resolver.sync(testFileDir, importPath);

	if (result.error || !result.path) {
		console.warn(
			`[oxc-resolver] Could not resolve "${importPath}" from "${testFileId}": ${result.error || "No path resolved"}. Using fallback resolution.`,
		);
		return fallbackResolveComponentPath(importPath, testFileId);
	}

	const projectRoot = process.cwd();
	const relativePath = relative(projectRoot, result.path);

	return relativePath.startsWith(".") ? relativePath : `./${relativePath}`;
}

export function hasCommandsImport(node: Node): boolean {
	if (
		!isImportDeclaration(node) ||
		node.source?.value !== "@vitest/browser/context" ||
		!node.specifiers
	) {
		return false;
	}

	return node.specifiers.some(
		(spec) =>
			spec.type === "ImportSpecifier" &&
			spec.imported.type === "Identifier" &&
			spec.imported.name === "commands",
	);
}

export async function renderComponentToSSR(
	ctx: BrowserCommandContext,
	Component: Component,
	props: Record<string, unknown> = {},
): Promise<{ html: string }> {
	const viteServer = ctx.project.vite;

	const qwikModule = await viteServer.ssrLoadModule("@builder.io/qwik");
	const { jsx } = qwikModule;
	const jsxElement = jsx(Component, props);

	const serverModule = await viteServer.ssrLoadModule(
		"@builder.io/qwik/server",
	);
	const { renderToString } = serverModule;

	const result = await renderToString(jsxElement, {
		containerTagName: "div",
		base: "/",
		qwikLoader: { include: "always" },
		symbolMapper: globalThis.qwikSymbolMapper,
	});

	return { html: result.html };
}
