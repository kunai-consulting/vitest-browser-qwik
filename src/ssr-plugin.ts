import { resolve } from "node:path";
import { symbolMapper } from "@builder.io/qwik/optimizer";
import type { Node } from "@oxc-project/types";
import MagicString from "magic-string";
import { parseSync } from "oxc-parser";
import type { Plugin } from "vitest/config";
import type { BrowserCommand } from "vitest/node";
import {
	extractPropsFromJSX,
	hasCommandsImport,
	hasRenderSSRCallInAST,
	isCallExpression,
	isExpressionStatement,
	isImportDeclaration,
	isJSXElement,
	isTestFile,
	isVariableDeclarator,
	renderComponentToSSR,
	resolveComponentPath,
	traverseChildren,
} from "./ssr-plugin-utils";

type ComponentFormat = BrowserCommand<
	[
		componentPath: string,
		componentName: string,
		props?: Record<string, unknown>,
	]
>;

type LocalComponentFormat = BrowserCommand<
	[
		testFilePath: string,
		componentName: string,
		allLocalComponents: string[],
		props?: Record<string, unknown>,
	]
>;

const renderSSRCommand: ComponentFormat = async (
	ctx,
	componentPath: string,
	componentName: string,
	props: Record<string, unknown> = {},
) => {
	const projectRoot = process.cwd();
	const absoluteComponentPath = resolve(projectRoot, componentPath);
	const viteServer = ctx.project.vite;

	// it does not replace env vars in the test file, so we need to do it manually
	if (!viteServer.config.define) return;
	for (const [key, value] of Object.entries(viteServer.config.env)) {
		viteServer.config.define[`__vite_ssr_import_meta__.env.${key}`] =
			JSON.stringify(value);
	}

	const componentModule = await viteServer.ssrLoadModule(absoluteComponentPath);
	const Component = componentModule[componentName];

	if (!Component) {
		throw new Error(
			`Component "${componentName}" not found in ${absoluteComponentPath}`,
		);
	}

	return await renderComponentToSSR(ctx, Component, props);
};

const renderSSRLocalCommand: LocalComponentFormat = async (
	ctx,
	testFilePath: string,
	componentName: string,
	allLocalComponents: string[],
	props: Record<string, unknown> = {},
) => {
	const viteServer = ctx.project.vite;

	// it does not replace env vars in the test file, so we need to do it manually
	if (!viteServer.config.define) return;
	for (const [key, value] of Object.entries(viteServer.config.env)) {
		viteServer.config.define[`__vite_ssr_import_meta__.env.${key}`] =
			JSON.stringify(value);
	}

	const { readFileSync, writeFileSync, unlinkSync } = await import("node:fs");
	const { dirname, join } = await import("node:path");

	const tempFileName = `ssr-test-${Date.now()}-${Math.random().toString(36).slice(2, 11)}.tsx`;
	// temp file inthe same folder to support relative imports
	const testFileDir = dirname(testFilePath);
	const tempFilePath = join(testFileDir, tempFileName);

	try {
		const originalContent = readFileSync(testFilePath, "utf8");
		const ast = parseSync(testFilePath, originalContent);
		const s = new MagicString(originalContent);

		function cleanTestFile(node: Node): undefined {
			if (isImportDeclaration(node)) {
				const source = node.source?.value;
				if (source === "vitest" || source?.includes("@vitest/")) {
					s.remove(node.start, node.end);
				}
			}

			if (
				isExpressionStatement(node) &&
				node.expression?.type === "CallExpression"
			) {
				const callExpr = node.expression;
				if (callExpr.callee.type === "Identifier") {
					const calleeName = callExpr.callee.name;
					if (
						calleeName === "test" ||
						calleeName === "describe" ||
						calleeName === "it"
					) {
						s.remove(node.start, node.end);
					}
				}
			}

			traverseChildren(node, cleanTestFile);
			return undefined;
		}

		cleanTestFile(ast.program);

		let cleanedContent = s.toString();
		if (allLocalComponents.length > 0) {
			const exportStatement = `\n\n// Auto-generated exports for local components\nexport { ${allLocalComponents.join(", ")} };`;
			cleanedContent += exportStatement;
		}

		writeFileSync(tempFilePath, cleanedContent, "utf8");

		const componentModule = await viteServer.ssrLoadModule(tempFilePath);
		const Component = componentModule[componentName];

		if (!Component) {
			throw new Error(
				`[vitest-browser-qwik]: Local component "${componentName}" not found in ${testFilePath}. Available exports: ${Object.keys(componentModule).join(", ")}`,
			);
		}

		return await renderComponentToSSR(ctx, Component, props);
	} finally {
		try {
			unlinkSync(tempFilePath);
		} catch (cleanupError) {
			console.warn("Failed to clean up temporary file:", cleanupError);
		}
	}
};

export function testSSR(): Plugin {
	return {
		name: "vitest:ssr-transform",
		enforce: "pre",

		async transform(code, id) {
			if (!isTestFile(id)) return null;

			const ast = parseSync(id, code);
			if (!hasRenderSSRCallInAST(ast.program, code)) return null;

			const s = new MagicString(code);
			const componentImports = new Map<string, string>();
			const localComponents = new Map<string, string>();
			const renderSSRIdentifiers = new Set<string>(["renderSSR"]);
			let hasExistingCommandsImport = false;

			function walkForTransformation(node: Node): undefined {
				if (
					isImportDeclaration(node) &&
					node.source?.value &&
					node.specifiers
				) {
					const source = node.source.value;
					for (const spec of node.specifiers) {
						if (
							spec.type === "ImportSpecifier" &&
							spec.imported.type === "Identifier"
						) {
							componentImports.set(spec.imported.name, source);
							if (spec.imported.name === "renderSSR") {
								renderSSRIdentifiers.add(spec.local.name);
							}
						} else if (
							spec.type === "ImportDefaultSpecifier" &&
							spec.local.name.toLowerCase().includes("renderssr")
						) {
							renderSSRIdentifiers.add(spec.local.name);
						}
					}
				}

				if (isVariableDeclarator(node)) {
					if (
						node.id.type === "Identifier" &&
						node.init?.type === "Identifier" &&
						renderSSRIdentifiers.has(node.init.name)
					) {
						renderSSRIdentifiers.add(node.id.name);
					}

					if (
						node.id.type === "Identifier" &&
						node.init?.type === "CallExpression"
					) {
						const callExpr = node.init;
						if (
							callExpr.callee.type === "Identifier" &&
							callExpr.callee.name === "component$"
						) {
							const fullDeclaration = code.slice(node.start, node.end);
							localComponents.set(node.id.name, fullDeclaration);
						}
					}
				}

				if (hasCommandsImport(node)) {
					hasExistingCommandsImport = true;
				}

				if (
					isCallExpression(node) &&
					node.callee.type === "Identifier" &&
					renderSSRIdentifiers.has(node.callee.name)
				) {
					const jsxArg = node.arguments?.[0];
					if (
						!isJSXElement(jsxArg) ||
						jsxArg.openingElement?.name?.type !== "JSXIdentifier"
					) {
						traverseChildren(node, walkForTransformation);
						return;
					}

					const componentName = jsxArg.openingElement.name.name;
					const props = extractPropsFromJSX(
						jsxArg.openingElement.attributes || [],
						code,
					);

					let propsStr = "";
					if (Object.keys(props).length > 0) {
						const propsEntries = Object.entries(props).map(
							([key, value]) => `${JSON.stringify(key)}: ${value}`,
						);
						propsStr = `, { ${propsEntries.join(", ")} }`;
					}

					const localComponentCode = localComponents.get(componentName);
					if (localComponentCode) {
						const allLocalComponentNames = Array.from(localComponents.keys());
						const localComponentsArray = JSON.stringify(allLocalComponentNames);
						const replacement = `(async () => {
							const { html } = await commands.renderSSRLocal("${id}", "${componentName}", ${localComponentsArray}${propsStr});
							return renderServerHTML(html);
						})()`;
						s.overwrite(node.start, node.end, replacement);
					} else {
						const componentImportPath = componentImports.get(componentName);
						if (componentImportPath) {
							const componentPath = resolveComponentPath(
								componentImportPath,
								id,
							);
							const replacement = `(async () => {
								const { html } = await commands.renderSSR("${componentPath}", "${componentName}"${propsStr});
								return renderServerHTML(html);
							})()`;
							s.overwrite(node.start, node.end, replacement);
						}
					}
				}

				traverseChildren(node, walkForTransformation);
				return undefined;
			}

			walkForTransformation(ast.program);

			if (s.hasChanged()) {
				if (!hasExistingCommandsImport) {
					let lastImportEnd = 0;

					function findLastImport(node: Node): undefined {
						if (isImportDeclaration(node)) {
							lastImportEnd = Math.max(lastImportEnd, node.end);
						}
						traverseChildren(node, findLastImport);
						return undefined;
					}

					findLastImport(ast.program);

					if (lastImportEnd > 0) {
						s.appendLeft(
							lastImportEnd,
							'\nimport { commands } from "@vitest/browser/context";\nimport { renderServerHTML } from "vitest-browser-qwik";',
						);
					}
				}

				if (localComponents.size > 0) {
					const localComponentNames = Array.from(localComponents.keys());
					const exportStatement = `\n\n// Auto-generated exports for local components\nexport { ${localComponentNames.join(", ")} };`;
					s.append(exportStatement);
				}

				return {
					code: s.toString(),
					map: s.generateMap({ hires: true }),
				};
			}

			return null;
		},
		configResolved(config) {
			globalThis.qwikSymbolMapper = symbolMapper;

			if (config.test?.browser?.enabled) {
				config.test.browser.commands = {
					...config.test.browser.commands,
					renderSSR: renderSSRCommand,
					renderSSRLocal: renderSSRLocalCommand,
				};
			}
		},
	};
}
