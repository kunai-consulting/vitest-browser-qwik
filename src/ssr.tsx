import type { JSXOutput } from "@builder.io/qwik";
import { renderToStream } from "@builder.io/qwik/server";
import type { Plugin } from "vitest/config";
import type { BrowserCommand } from "vitest/node";

const renderSSR: BrowserCommand<[component: JSXOutput]> = async (component) => {
	// Set up environment variables that Qwik SSR expects
	if (!process.env.BASE_URL) {
		process.env.BASE_URL = "/";
	}
	if (!process.env.VITE_BASE_URL) {
		process.env.VITE_BASE_URL = "/";
	}

	let html = "";

	const stream = {
		write: (chunk: string) => {
			html += chunk;
		},
	};

	await renderToStream(component, {
		containerTagName: "div",
		stream,
		base: "/build/",
	});

	return {
		html,
	};
};

export default function SSRCommand(): Plugin {
	return {
		name: "vitest:ssr-command",
		config() {
			return {
				test: {
					browser: {
						commands: {
							renderSSR,
						},
					},
				},
			};
		},
	};
}
