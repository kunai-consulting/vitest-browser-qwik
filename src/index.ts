import type { JSXOutput } from "@builder.io/qwik";
import { page } from "@vitest/browser/context";
import { beforeEach } from "vitest";
import { cleanup, render, renderServerHTML } from "./pure";

/** This is replaced with actual code by the ssr-plugin.ts transform */
export declare function renderSSR(
	jsxNode: JSXOutput,
): Promise<import("./pure").RenderResult>;

export type { RenderResult, SSRRenderOptions } from "./pure";
export {
	cleanup,
	render,
	renderHook,
	renderServerHTML,
} from "./pure";

page.extend({
	render,
	renderServerHTML: renderServerHTML,
	[Symbol.for("vitest:component-cleanup")]: cleanup,
});

beforeEach(() => {
	cleanup();
});

declare module "@vitest/browser/context" {
	interface BrowserPage {
		render: typeof render;
		renderServerHTML: typeof renderServerHTML;
	}
}
