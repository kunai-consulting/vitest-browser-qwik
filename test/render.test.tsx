import { page, userEvent } from "@vitest/browser/context";
// import { Button } from "react-aria-components";
import { expect, test, vi } from "vitest";
import { render } from "../src/index";
import { Counter } from "./fixtures/Counter";
import { HelloWorld } from "./fixtures/HelloWorld";

test("renders simple component", async () => {
	const screen = render(<HelloWorld />);
	await expect.element(page.getByText("Hello World")).toBeVisible();
	expect(screen.container).toMatchSnapshot();
});

// test("renders counter", async () => {
// 	const screen = render(<Counter initialCount={1} />);

// 	// await expect.element(screen.getByText("Count is 1")).toBeVisible();
// 	// await screen.getByRole("button", { name: "Increment" }).click();
// 	// await expect.element(screen.getByText("Count is 2")).toBeVisible();
// });

// test("should fire the onPress/onClick handler", async () => {
// 	const handler = vi.fn();
// 	const screen = page.render(<Button onPress={handler}>Button</Button>);
// 	await userEvent.click(screen.getByRole("button"));
// 	// await screen.getByRole('button').click()
// 	expect(handler).toHaveBeenCalled();
// });
