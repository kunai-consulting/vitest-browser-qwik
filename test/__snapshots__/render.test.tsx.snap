// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders simple component 1`] = `
<div
  q:container="resumed"
  q:render="dom"
  q:version="1.17.0"
>
  <!--qv -->
  <div>
    Hello World
  </div>
  <!--/qv-->
</div>
`;
